crates/ibc/src/clients/ics07_tendermint/client_state/misbehaviour.rs

```diff
@@ -0,0 +1,163 @@
+use crate::core::{ics24_host::identifier::ClientId, ValidationContext};
+use crate::timestamp::Timestamp;
+
+use super::{check_header_trusted_next_validator_set, downcast_tm_consensus_state, ClientState};
+
+impl ClientState {
+    // verify_misbehaviour determines whether or not two conflicting headers at
+    // the same height would have convinced the light client.
+    pub fn verify_misbehaviour(
+        &self,
+        ctx: &dyn ValidationContext,
+        client_id: &ClientId,
+        misbehaviour: TmMisbehaviour,
+    ) -> Result<(), ClientError> {
+        let header_1 = misbehaviour.header1();
+        let trusted_consensus_state_1 = {
+            let consensus_state_path =
+                ClientConsensusStatePath::new(client_id, &header_1.trusted_height);
+            let consensus_state = ctx.consensus_state(&consensus_state_path)?;
+
+            downcast_tm_consensus_state(consensus_state.as_ref())
+        }?;
+
+        let header_2 = misbehaviour.header2();
+        let trusted_consensus_state_2 = {
+            let consensus_state_path =
+                ClientConsensusStatePath::new(client_id, &header_2.trusted_height);
+            let consensus_state = ctx.consensus_state(&consensus_state_path)?;
+
+            downcast_tm_consensus_state(consensus_state.as_ref())
+        }?;
+
+        self.check_misbehaviour_headers_consistency(header_1, header_2)?;
+
+        let current_timestamp = ctx.host_timestamp()?;
+        self.check_misbehaviour_header(header_1, &trusted_consensus_state_1, current_timestamp)?;
+        self.check_misbehaviour_header(header_2, &trusted_consensus_state_2, current_timestamp)
+    }
+
+    pub fn check_misbehaviour_headers_consistency(
+        &self,
+        header_1: &TmHeader,
+        header_2: &TmHeader,
+    ) -> Result<(), ClientError> {
+        if header_1.signed_header.header.chain_id != header_2.signed_header.header.chain_id {
+            return Err(Error::InvalidRawMisbehaviour {
+                reason: "headers must have identical chain_ids".to_owned(),
+            }
+            .into());
+        }
+
+        if header_1.height() < header_2.height() {
+            return Err(Error::InvalidRawMisbehaviour {
+                reason: format!(
+                    "headers1 height is less than header2 height ({} < {})",
+                    header_1.height(),
+                    header_2.height()
+                ),
+            }
+            .into());
+        }
+
+        Ok(())
+    }
+
+    pub fn check_misbehaviour_header(
+        &self,
+        header: &TmHeader,
+        trusted_consensus_state: &TmConsensusState,
+        current_timestamp: Timestamp,
+    ) -> Result<(), ClientError> {
+        // ensure correctness of the trusted next validator set provided by the relayer
+        check_header_trusted_next_validator_set(header, trusted_consensus_state)?;
+
+        // ensure header timestamp is within trusted period from the trusted consensus state
```

Comment:
The code is correct but the comment is not. We could be receiving old headers still within the trusted ~~consensus state~~ period while the trusted consensus state is expired. 
```suggestion
        // ensure trusted consensus state is within trusting period
```

Reply:
hmm, not sure I follow. The model that I have is

```
(1)                       (2)                        (3) end of trusting period
+--------------------------+--------------------------+
trusted CS             new header 
```

My comment was trying to express: `(1) <= (2) <= (3)`. IIUC, your comment takes a different but equivalent perspective:

```
(1)                       (2)                        (3) new header
+--------------------------+--------------------------+
end of trusting period    trusted CS 
```
That is you look at the trusting period as "going back" from `now` (3) whereas I see it as "going forward" from the trusted CS.

Is this correct? Then I don't understand

> while the trusted consensus state is expired.

what does it mean for a trusted consensus state to be expired?

Reply:
```
(1)                       (2)                        (3) current_timestamp
+--------------------------+--------------------------+
trusted CS             new header 
```
the comment says `(2) - (1) < trusting_period` but the code does `(3) - (1) < trusting_period`
Note: `(2) - (1) < trusting_period` check is not correct as it may still pass if we update with an older header while trusted cs is expired.

> what does it mean for a trusted consensus state to be expired?

it means it cannot be trusted anymore.

Reply:
ah, yes, get it now